Project 4_KTs

```diff
@@ -0,0 +1,446 @@
+
+The query should return the name, address, and average rating of each restaurant. The restaurants should be ordered in descending order based on their average rating.
+
+Tags: SPARQL, restaurants, Pho, Buffalo, ratings, Grubhub, UberEats, DoorDash, Yelp
+
+Example:
+
+Consider the following RDF data (Turtle):
+
+@prefix schema: <http://schema.org/> .
+@prefix ex: <http://example.org/> .
+
+ex:restaurant1 a schema:Restaurant ;
+    schema:name "Pho 99" ;
+    schema:address "357 Connecticut St, Buffalo, NY 14213" ;
+    schema:aggregateRating [
+        a schema:AggregateRating ;
+        schema:ratingValue "4.2" ;
+        schema:reviewCount "75" ;
+        schema:reviewRating [
+            a schema:Rating ;
+            schema:author "Grubhub" ;
+            schema:ratingValue "4.2"
+        ]
+    ].
+
+ex:restaurant2 a schema:Restaurant ;
+    schema:name "Pho Dollar" ;
+    schema:address "322 W Ferry St, Buffalo, NY 14213" ;
+    schema:aggregateRating [
+        a schema:AggregateRating ;
+        schema:ratingValue "4.5" ;
+        schema:reviewCount "45" ;
+        schema:reviewRating [
+            a schema:Rating ;
+            schema:author "UberEats" ;
+            schema:ratingValue "4.5"
+        ]
+    ].
+
+ex:restaurant3 a schema:Restaurant ;
+    schema:name "Pho 54" ;
+    schema:address "128 Elmwood Ave, Buffalo, NY 14201" ;
+    schema:aggregateRating [
+        a schema:AggregateRating ;
+        schema:ratingValue "4.8" ;
+        schema:reviewCount "65" ;
+        schema:reviewRating [
+            a schema:Rating ;
+            schema:author "DoorDash" ;
+            schema:ratingValue "4.8"
+        ]
+    ].
+
+ex:restaurant4 a schema:Restaurant ;
+    schema:name "Pho Lantern" ;
+    schema:address "837 Niagara St, Buffalo, NY 14213" ;
+    schema:aggregateRating [
+        a schema:AggregateRating ;
+        schema:ratingValue "4.7" ;
+        schema:reviewCount "90" ;
+        schema:reviewRating [
+            a schema:Rating ;
+            schema:author "Yelp" ;
+            schema:ratingValue "4.7"
+        ]
+    ].
+
+The following query should return the best Pho restaurants in Buffalo, NY based on ratings from multiple platforms (SPARQL):
+
+---
+PREFIX schema: <http://schema.org/>
+SELECT ?name ?address (AVG(?ratingValue) AS ?averageRating)
+WHERE {
+    ?restaurant a schema:Restaurant ;
+                schema:name ?name ;
+                schema:address ?address ;
+                schema:aggregateRating ?aggregateRating .
+
+    ?aggregateRating schema:ratingValue ?ratingValue ;
+                      schema:reviewRating ?reviewRating .
+
+    ?reviewRating schema:author ?author .
+
+    FILTER (REGEX(?address, "Buffalo, NY") && (?author = "Grubhub" || ?author = "UberEats" || ?author = "DoorDash" || ?author = "Yelp"))
+}
+GROUP BY ?name ?address
+ORDER BY DESC(?averageRating)
+---
+
```

Comment:
I love this. This is a much more complicated problem and solution.

I think this should be a Kata-5, or Kata-4 at most. See my prior comments on what I would expect for a Kata at level 4. What I see is really good here, is
- generating a new value after a calculation (MAX)
- queries' triples are chained, even if it could be more exotic
- filter with a regular-expression (regex) is difficult, but comprehensible. Though, the syntax could be to have the disjunctions a single-stroke ( | , rather than double, || ), and all of the strings of author name gathered together (e.g., "Grubhub"|"UberEats"|"DoorDash"|"Yelp" ... Unless I have misunderstood the regex syntax)
- manipulating data in two different ways (group by, order by), though I think the syntax is a little off. (Doesn't 'DESC' come after the variable? Or is that just SQL?)

Reply:
RE: the last bullet -- the "DESC" comes before the variable in SPARQL, but after in SQL. Sorry for the confusion.